ext/liquid_c/block.c

```diff
@@ -123,65 +210,199 @@ static VALUE internal_block_parse(VALUE self, VALUE tokens, VALUE parse_context,
+        .ruby_obj = parse_context_obj,
+    };
+    Tokenizer_Get_Struct(tokenizer_obj, parse_context.tokenizer);
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+
+    ensure_not_parsing(body);
+    if (body->source == Qnil) {
+        body->source = parse_context.tokenizer->source;
+    } else if (body->source != parse_context.tokenizer->source) {
+        rb_raise(rb_eArgError, "Liquid::C::BlockBody#parse must be passed the same tokenizer when called multiple times");
+    }
+    body->parsing = true;
+    // remove terminating OP_LEAVE to resume parsing
+    body->instructions.size -= 1;
+
+    tag_markup_t unknown_tag = internal_block_body_parse(body, &parse_context);
+    block_body_add_leave(body);
+    body->parsing = false;
+    return rb_yield_values(2, unknown_tag.name, unknown_tag.markup);
+}
+
+static VALUE block_body_render_to_output_buffer(VALUE self, VALUE context, VALUE output)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+    liquid_vm_render(body, context, output);
+    return output;
+}
 
-                rb_ary_push(nodelist, new_tag);
+static VALUE block_body_blank_p(VALUE self)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+    return body->blank ? Qtrue : Qfalse;
+}
+
+static VALUE block_body_remove_blank_strings(VALUE self)
+{
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+
+    if (!body->blank) {
+        rb_raise(rb_eRuntimeError, "remove_blank_strings only support being called on a blank block body");
+    }
+    ensure_not_parsing(body);
+
+    size_t *const_ptr = (size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+
+    while (true) {
+        switch (*ip++) {
+            case OP_LEAVE:
+                goto loop_break;
+            case OP_WRITE_RAW:
+            {
+                const_ptr++;
+                size_t *size_ptr = const_ptr++;
+                if (*size_ptr) {
+                    *size_ptr = 0; // effectively a no-op
+                    body->render_score--;
+                }
                 break;
             }
+            case OP_WRITE_NODE:
+                const_ptr++;
+                break;
+            default:
+                rb_bug("invalid opcode: %u", ip[-1]);
         }
     }
+loop_break:
+
     return Qnil;
 }
 
-static VALUE rb_block_parse(VALUE self, VALUE tokens, VALUE parse_context)
+// Deprecated: avoid using this for the love of performance
+static VALUE block_body_nodelist(VALUE self)
 {
-    return internal_block_parse(self, tokens, parse_context, false);
+    block_body_t *body;
+    BlockBody_Get_Struct(self, body);
+
+    ensure_not_parsing(body);
+
+    // Use rb_attr_get insteaad of rb_ivar_get to avoid an instance variable not
+    // initialized warning.
+    VALUE nodelist = rb_attr_get(self, intern_ivar_nodelist);
+    if (nodelist != Qnil)
+        return nodelist;
+    nodelist = rb_ary_new_capa(body->instructions.size / sizeof(VALUE));
```

Comment:
Since this is just the capacity it doesn't matter, but it looks like this is including an extra index for the OP_LEAVE.